### PR TITLE
Urban area mapping using Sentinel 1 notebook

### PR DESCRIPTION
**Pull request by vikineema**
### Proposed changes
Add the Urban area mapping using Sentinel 1 notebook to the Real World examples. 

### Closes issues (optional)
- Closes Issue #000
- Closes Issue #000

### Checklist (replace `[ ]` with `[x]` to check off)
- [x] Remove any unused Python packages from `Load packages`
- [x] Remove any unused/empty code cells
- [x] Remove any guidance cells (e.g. `General advice`)
- [x] Ensure that all code cells follow the [PEP8 standard](https://www.python.org/dev/peps/pep-0008/) for code. The `jupyterlab_code_formatter` tool can be used to format code cells to a consistent style: select each code cell, then click `Edit` and then one of the `Apply X Formatter` options (`YAPF` or `Black` are recommended).
- [x] Include relevant tags in the final notebook cell and re-use tags if possible
- [x] Clear all outputs, run notebook from start to finish, and save the notebook in the state where all cells have been sequentially evaluated

### CODE REVIEW

**Comment by cbur24** [comment]
Very intersesting notebook, @vikineema! A few points to consider before we can merge into the repo
1. If you're going to resample S1 into an annual median, there's no need to do speckle/lee filtering.
2. If you're resampling S1 to annual, why not just use the ALOS/PALSAR annual mosaics? Then you can simple load the annual mosaic and save doing all that unnecessary compute. We have annual mosaic up to 2020 https://maps.digitalearth.africa/#share=s-4ajkduucUi9kGhVWd3Ifzg1Didi
3. I don't think the notebook should explore so many different methods for classifying, the notebook will be more succinct and useful if you can identify the best method and only show that.  
4. There should be a better (simpler/cleaner) way to compare the results between your classifications and the LULC maps, scikit-learn has methods for accuracy assessment that you can use instead of defining those long functions.
5. Why are you defining two new functions `raster_stack` and `raster_unstack` instead of using `deafrica_tools.classification.sklearn_flatten` and `deafrica_tools.classification.sklearn_unflatten`? If you have methods for improving those functions then the PR should include changes to those functions, rather than defining them in the notebook. And the PR should explain why you've made these edits.
6.  It might be worth exploring if you get better results if you include bands from another dataset like `gm_s2_annual`. You could load the S2 geomad and ALOS annual mosaic and use those datasets for classifying urban areas with the kmean clustering method.

And lastly, when you are ready for this PR to be reviewed for merging into the `main` branch, change the PR from draft to 'ready'

**Comment by eefaye** [comment]
@cbur24 so far we have been treating the `deafrica_tools.classification` rework as a separate item (as you have noticed, one of the sticking points is the `stack/unstack` vs `flatten/unflatten` functions). I agree it makes sense to push that first before or as part of this PR. @vikineema if you think it will be extensive, the `classification` rework might be better as a separate PR? We can discuss this tomorrow.

@vikineema I think this is a very thorough analysis, looks really good. It is pretty long for a RWE notebook though. We could shape it more by deciding the angle/audience of the notebook. Do we want it to focus on:
 - machine learning?
 - that you should try different analysis methods?

_Machine learning:_
If this is the focus, then the Otsu thresholding could move out, perhaps into its own notebook. Could go straight into Kmeans and then it might be possible to add extra bands like Chad suggested (although that might just make it even longer).
_Different analysis methods:_
More like it is now, which would pair nicely with the [urban index comparison notebook](https://github.com/digitalearthafrica/deafrica-sandbox-notebooks/blob/main/Real_world_examples/Urban_index_comparison.ipynb) as that compares indices and this compares methods. The notebook can have a central plot of "how do the methods compare and which one is better?" Some of the explanations could be condensed and I would suggest only selecting either `vh` or `vv` to threshold, not both. The notebook could conclude with a quick look at relative strengths/weaknesses of each method.

More general comments:
 - `try_all_threshold` is great for experimenting but probably doesn't need to be in a Real World Example notebook. It can be removed from here. 
 - `In[34]-[36]` might work better as one plot with subplots, to make it easier for the reader to compare results

_Just a thought:_ If the speckle filter is something which we think has longevity (I have no experience in this @cbur24 ? @fangfy ?) we could put it straight in `deafrica_tools.datahandling` and integrate it into the [Sentinel-1 dataset](https://github.com/digitalearthafrica/deafrica-sandbox-notebooks/blob/main/Datasets/Sentinel_1.ipynb) notebook. Is it correct to say we currently show how to mask S-1 data but filtering is left as an exercise to the reader? Are there existing speckle filter packages that work on our data?

**Comment by vikineema** [comment]
@cbur24, @eefaye 
In the recent updates I have made the following changes:
1. I removed the speckle filtering.
2. For the Kmeans clustering I used the S1 annual median instead of the ALOS/PALSAR annual mosaic for 2020. This is because when I tested the two products I got an accuracy of 74.06 using the ALOS/PALSAR annual mosaic which is lower than the accuracy of  79.972 that I got from using the S1 annual median. 
3. In the radar urban area mapping notebook, I used kmeans clustering.  For the thresholding I put it into the frequently used code.
4. Used the sklearn metrics functions to compare the results between the kmeans classification and the LULC maps.
5. I have improved the sklearn_flatten and unflatten functions to improve the speed of the functions as seen in the test notebook in: https://github.com/digitalearthafrica/deafrica-sandbox-notebooks/blob/test-classification-tools/Real_world_examples/Urban_area_mapping/Timing_functions_radar_urban_area_mapping.ipynb

**Comment by vikineema** [comment]
@cbur24 @eefaye In the recent changes I have made, I have reverted the sklearn_flatten and unflatten to the original functions and run the classification using the vv band.

**Comment by caitlinadams** [comment]
I just ran this, and it looks like the class values for k-means might be changing. In this notebook, it says class 1 in k=3 should be the urban area, but looking at the plot I got, I think it should actually be class 2. We might need to find a way to make this more reproducible.

Screen Shot 2021-12-22 at 3 33 56 pm

**Review by caitlinadams — CHANGES_REQUESTED** [review]
We'll need to find a way to fix the class value for the kmeans approach to make the notebook reproducible. See my comment above.

**Comment by eefaye** [comment]
I think putting the `random_state` argument in the `kmeans_clustering` function fixes the outcome. I've set `random_state=1` and have run it a few times but would appreciate confirmation that it works for other people.

**Comment by caitlinadams** [comment]
Closing now that the updated PR has been merged.
